File: 0501-find-mode-in-binary-search-tree/0501-find-mode-in-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    Map<Integer, Integer> map = new HashMap<>();
    public int[] findMode(TreeNode root) {
        preOrder(root);
        int maxFreq = 0;
        for(int key : map.keySet()){
            maxFreq = Math.max(maxFreq, map.get(key));
        }
        List<Integer> list = new ArrayList<>();
        for(int key : map.keySet()){
            if(map.get(key) == maxFreq) list.add(key);
        }
        int[] res = list.stream().mapToInt(Integer::intValue).toArray();
        return res;
    }
    public void preOrder(TreeNode root){
        if(root == null) return;
        map.put(root.val, map.getOrDefault(root.val, 0) + 1);
        preOrder(root.left);
        preOrder(root.right);
    }
}

File: 0501-find-mode-in-binary-search-tree/0501-find-mode-in-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def findMode(self, root: Optional[TreeNode]) -> List[int]:
        freq_map = defaultdict(int)
        def pre_order(node):
            if not node:
                return
            freq_map[node.val] += 1
            pre_order(node.left)
            pre_order(node.right)
        pre_order(root)
        max_freq = max(freq_map.values())
        return [x for x in freq_map if freq_map[x] == max_freq]
        
        
        